### ISSUE TITLE
v1.22.1 breaks VCR specs

### ISSUE THREAD

**Issue by baburdick** [bblimke/webmock#536]
webmock v1.22.1 breaks our VCR specs. Downgrading to v1.22.0 resolves the issue.
```
Failures:

  1) MorningCampaignCleanup#call should remove all existing lists from campaign
     Failure/Error: response = interactor.cleanup_old_campaign_lists(existing_lists)
     VCR::Errors::UnhandledHTTPRequestError:
       
       
       ================================================================================
       An HTTP request has been made that VCR does not know how to handle:
         GET https://some_user%40domain.net:some_password@api.vendor.com/wsadmin/v3/AdminWebService?user=some_user@domain.net&wsdl=true
       
       VCR is currently using the following cassette:
         - ~/Code/Domain/my_service/spec/fixtures/remove_lists_from_campaign.yml
         - :record => :once
         - :match_requests_on => [:method, :uri]
       
       Under the current configuration VCR can not find a suitable HTTP interaction
       to replay and is prevented from recording new requests. There are a few ways
       you can deal with this:
       
         * If you're surprised VCR is raising this error
           and want insight about how VCR attempted to handle the request,
           you can use the debug_logger configuration option to log more details [1].
         * You can use the :new_episodes record mode to allow VCR to
           record this new request to the existing cassette [2].
         * If you want VCR to ignore this request (and others like it), you can
           set an `ignore_request` callback [3].
         * The current record mode (:once) does not allow new requests to be recorded
           to a previously recorded cassette. You can delete the cassette file and re-run
           your tests to allow the cassette to be recorded with this request [4].
         * The cassette contains 2 HTTP interactions that have not been
           played back. If your request is non-deterministic, you may need to
           change your :match_requests_on cassette option to be more lenient
           or use a custom request matcher to allow it to match [5].
       
       [1] https://www.relishapp.com/vcr/vcr/v/2-9-3/docs/configuration/debug-logging
       [2] https://www.relishapp.com/vcr/vcr/v/2-9-3/docs/record-modes/new-episodes
       [3] https://www.relishapp.com/vcr/vcr/v/2-9-3/docs/configuration/ignore-request
       [4] https://www.relishapp.com/vcr/vcr/v/2-9-3/docs/record-modes/once
       [5] https://www.relishapp.com/vcr/vcr/v/2-9-3/docs/request-matching
       ================================================================================
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/vcr-2.9.3/lib/vcr/request_handler.rb:97:in `on_unhandled_request'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/vcr-2.9.3/lib/vcr/library_hooks/webmock.rb:112:in `on_unhandled_request'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/vcr-2.9.3/lib/vcr/request_handler.rb:24:in `handle'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/vcr-2.9.3/lib/vcr/library_hooks/webmock.rb:127:in `block in <module:WebMock>'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/webmock-1.22.1/lib/webmock/stub_registry.rb:28:in `call'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/webmock-1.22.1/lib/webmock/stub_registry.rb:28:in `block in register_global_stub'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/webmock-1.22.1/lib/webmock/request_pattern.rb:36:in `call'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/webmock-1.22.1/lib/webmock/request_pattern.rb:36:in `matches?'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/webmock-1.22.1/lib/webmock/stub_registry.rb:58:in `block in request_stub_for'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/webmock-1.22.1/lib/webmock/stub_registry.rb:57:in `each'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/webmock-1.22.1/lib/webmock/stub_registry.rb:57:in `detect'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/webmock-1.22.1/lib/webmock/stub_registry.rb:57:in `request_stub_for'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/webmock-1.22.1/lib/webmock/stub_registry.rb:50:in `response_for_request'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/webmock-1.22.1/lib/webmock/http_lib_adapters/net_http.rb:79:in `request'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/httpi-2.1.1/lib/httpi/adapter/net_http.rb:55:in `perform'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/httpi-2.1.1/lib/httpi/adapter/net_http.rb:42:in `block in request'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/httpi-2.1.1/lib/httpi/adapter/net_http.rb:68:in `call'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/httpi-2.1.1/lib/httpi/adapter/net_http.rb:68:in `block in do_request'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/webmock-1.22.1/lib/webmock/http_lib_adapters/net_http.rb:123:in `start_without_connect'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/webmock-1.22.1/lib/webmock/http_lib_adapters/net_http.rb:150:in `start'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/httpi-2.1.1/lib/httpi/adapter/net_http.rb:66:in `do_request'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/httpi-2.1.1/lib/httpi/adapter/net_http.rb:33:in `request'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/httpi-2.1.1/lib/httpi.rb:140:in `request'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/httpi-2.1.1/lib/httpi.rb:106:in `get'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/wasabi-3.3.1/lib/wasabi/resolver.rb:43:in `load_from_remote'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/wasabi-3.3.1/lib/wasabi/resolver.rb:33:in `resolve'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/wasabi-3.3.1/lib/wasabi/document.rb:142:in `xml'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/wasabi-3.3.1/lib/wasabi/document.rb:160:in `parse'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/wasabi-3.3.1/lib/wasabi/document.rb:147:in `parser'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/wasabi-3.3.1/lib/wasabi/document.rb:64:in `soap_actions'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/savon-2.5.1/lib/savon/operation.rb:21:in `ensure_exists!'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/savon-2.5.1/lib/savon/operation.rb:14:in `create'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/savon-2.5.1/lib/savon/client.rb:32:in `operation'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/savon-2.5.1/lib/savon/client.rb:36:in `call'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/five9_cws-0.2.0/lib/five9_cws/client.rb:33:in `make_call'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/five9_cws-0.2.0/lib/five9_cws/campaign_client.rb:50:in `remove_lists_from_campaign'
     # ./app/interactors/morning_campaign_cleanup.rb:153:in `block in cleanup_old_campaign_lists'
     # ./app/interactors/morning_campaign_cleanup.rb:152:in `each'
     # ./app/interactors/morning_campaign_cleanup.rb:152:in `cleanup_old_campaign_lists'
     # ./spec/interactors/morning_campaign_cleanup_spec.rb:50:in `block (4 levels) in <top (required)>'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/vcr-2.9.3/lib/vcr/util/variable_args_block_caller.rb:9:in `call'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/vcr-2.9.3/lib/vcr/util/variable_args_block_caller.rb:9:in `call_block'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/vcr-2.9.3/lib/vcr.rb:182:in `use_cassette'
     # ./spec/interactors/morning_campaign_cleanup_spec.rb:49:in `block (3 levels) in <top (required)>'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:206:in `instance_exec'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:206:in `block in run'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:430:in `block in with_around_and_singleton_context_hooks'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:388:in `block in with_around_example_hooks'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/hooks.rb:478:in `block in run'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/hooks.rb:618:in `block in run_around_example_hooks_for'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:273:in `call'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:273:in `call'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-rails-3.3.3/lib/rspec/rails/adapters.rb:127:in `block (2 levels) in <module:MinitestLifecycleAdapter>'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:378:in `instance_exec'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:378:in `instance_exec'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/hooks.rb:389:in `execute_with'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/hooks.rb:620:in `block (2 levels) in run_around_example_hooks_for'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:273:in `call'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:273:in `call'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/hooks.rb:621:in `run_around_example_hooks_for'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/hooks.rb:478:in `run'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:388:in `with_around_example_hooks'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:430:in `with_around_and_singleton_context_hooks'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:203:in `run'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example_group.rb:559:in `block in run_examples'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example_group.rb:555:in `map'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example_group.rb:555:in `run_examples'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example_group.rb:521:in `run'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example_group.rb:522:in `block in run'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example_group.rb:522:in `map'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/example_group.rb:522:in `run'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/runner.rb:115:in `block (3 levels) in run_specs'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/runner.rb:115:in `map'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/runner.rb:115:in `block (2 levels) in run_specs'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/configuration.rb:1627:in `with_suite_hooks'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/runner.rb:114:in `block in run_specs'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/reporter.rb:77:in `report'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/runner.rb:113:in `run_specs'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/runner.rb:89:in `run'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/runner.rb:73:in `run'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/lib/rspec/core/runner.rb:41:in `invoke'
     # ~/.rvm/gems/ruby-2.1.2@my_service/gems/rspec-core-3.3.2/exe/rspec:4:in `<main>'
```

**Comment by davidbegin** [comment]
Thanks for submitting this issue!

So if 1.22.1 doesn't work for you, 1.22.0 shouldn't either.
Because 1.22.0 was actually a bad release we pulled, on account we required a rack/utils, when "rack" isn't part of our dependencies.

The change is very small and you can see if here
https://github.com/bblimke/webmock/commit/b18b029ab9e4f9556cae590c1cb72beee54ae853

Are you sure 1.22.0 works for you, or do you mean 1.21?

If you are indeed using 1.22.0, it has to have something with requiring
rack/utils.  If you could include a small example reproducing this bug I could take a closer look, and try and find out if it is an issue on our end.

**Comment by baburdick** [comment]
You are correct. I mis-typed it. I'm not using v1.22.0. I'm using v1.21.0.

**Comment by davidbegin** [comment]
Ok, that makes a lot more sense, but I am still not sure where it could be!

But from looking through the [Changelog](https://github.com/bblimke/webmock/blob/master/CHANGELOG.md)
I am guessing it has to do with one of the following changes:

This one is specific to Net HTTP, which it looks like Httpi is using from the Stacktrace.
https://github.com/bblimke/webmock/commit/8a6f58e7b9fda7b72d15b55fcc5d95554492576b

This is around converting boolean's to strings, and I see you have true at the end of your URL
https://github.com/bblimke/webmock/commit/edb2245ba76974cecb07a2bff0923721e40ca131

Maybe you could Git Bisect to triagulate the commit that broke your spec.

Or with failing spec I can dig in and take a harder look.

**Comment by bblimke** [comment]
I can see `&wsdl=true` in the url. Any change your Faraday was upgraded together with WebMock and it's this issue? https://github.com/lostisland/faraday/issues/536

**Comment by dhague** [comment]
I also have this problem moving from 1.21 to 1.22.1 - in our case we are issuing a request `http://devstack.openstack.stack:35357/v3/projects/7720c2b0486e4b4e8f98605f1a3a2174?subtree_as_ids` and webmock is adding `=true` to the end.

**Comment by bblimke** [comment]
@dhague has any other gem been upgraded together with WebMock? Faraday perhaps?

**Comment by dhague** [comment]
No - the only differences between the working and non-working Gemfile.lock files are webmock (1.21.0 to 1.22.1) and hashdiff (non-existent to 0.2.2)

**Comment by bblimke** [comment]
What http client do you use?

**Comment by dhague** [comment]
We use Excon and VCR in this case - you can see the full stack trace of a failure here: https://travis-ci.org/dhague/fog/jobs/87058862#L343

**Comment by bblimke** [comment]
@davidbegin has anything been changed in 1.22.1 that't adding 'true' value to params with no value?

**Comment by davidbegin** [comment]
@bblimke I think it might have something to do with this commit
https://github.com/bblimke/webmock/commit/edb2245ba76974cecb07a2bff0923721e40ca131

I'll try and write up a spec to isolate this change in behavior tonight.

**Comment by davidbegin** [comment]
This is the commit that that triggered this change in behavior https://github.com/bblimke/webmock/commit/edb2245ba76974cecb07a2bff0923721e40ca131

**Comment by davidbegin** [comment]
The real issue was transforming nil into true, before converting params to string query params here: https://github.com/bblimke/webmock/blob/master/lib/webmock/util/query_mapper.rb#L81-L85

I'm still exploring why this logic was originally put in place.

**Comment by davidbegin** [comment]
I opened a Pull Request with my first attempt at fixing the issue here https://github.com/bblimke/webmock/pull/538

**Comment by bblimke** [comment]
The @davidbegin 's fix is now released as 1.22.2

**Comment by baburdick** [comment]
Confirmed that v1.22.2 works. Thanks!
